New: Add new props for polyglot options

soda0289 opened the pull request:
This commits adds new props for the following polylgot options:
   - allowMissing
   - onMissingKey
   - interpolation

 It also removes the use of componentWillReceiveProps() since it has
 been deprecated and replaces it with componentDidUpdate().

bastienrobert commented on src/i18n.d.ts:
```diff
@@ -2,13 +2,21 @@
+  interpolation?: {
+    suffix: string,
+    prefix: string
+  },
```
What about: 

```
interface InterpolationOptions {
  smart_count?: number | { length: number }
  _?: string

  [interpolationKey: string]: any
}

interface InterpolationTokenOptions {
  prefix?: string
  suffix?: string
}

interface I18nProps {
  children: ReactNode
  /** Locale to use, e.g. `en` */
  locale: number
  /** A dictionary of translations */
  messages: object
  /** control whether missing keys in a t call are allowed */
  allowMissing?: boolean;
  /** called instead of the default functionality if allowMissing */
  onMissingKey?: (key: string, options?: Polyglot.InterpolationOptions, locale?: string) => string
  /** change the substitution syntax for interpolation */
  interpolation?: InterpolationTokenOptions
}
```

soda0289 replied:
Thanks for the review.

Are you suggesting we don't use `InterpolationOptions`  from node-polyglot?

I think locale type should be `string | number` since even the example is using 'en'. Will make this change and push again

bastienrobert replied:
I absolutely agree with you about `string | number` for locale option

`InterpolationOptions` could be used from node-polyglot types but is I thought that DefinitelyTyped packages usually took some time to follow updates. Perhaps it would have been easier to directly control all types of the API internally. But I agree, maintaining the node-polyglot types in parallel to maintain this library is sooo much cleaner

bastienrobert replied:
Oh and maybe we could add `pluralRules` options too? This option is also not typed on DefinitelyTyped for example :(

nayaabkhan replied:
Totally agree on the point to contribute to typings for `node-polyglot` and use the types from there 👍

bastienrobert commented:
@nayaabkhan is it possible to merge it ASAP? It seems to be a bit urgent for some people (cf. #31) 🙃

nayaabkhan commented:
@soda0289 Thank you for the PR :)
@bastienrobert Be sure that I will take a look at this soon, it is on my priority list. BTW love the urgency :)

soda0289 commented:
@bastienrobert Pushed new fix.

Added in pluralRules. Did not test as I'm not sure what the option does but can look into later this weekend.

Also removed the use of node-polyglot types and created interfaces for InterpolationOptions and InterpolationTokenOptions.

Let me know if I should change anything else.

nayaabkhan commented on src/i18n.d.ts:
```diff
@@ -3,12 +3,37 @@
 
 import { ComponentType, ReactNode } from 'react'
 
+interface InterpolationOptions {
```
Is there any good reason not to rather `import { InterpolationOptions, InterpolationTokenOptions, PolyglotOptions } from 'node-polyglot'`?

soda0289 replied:
Node-polyglot doesn't have official types, its just using the @types/node-polyglot package from DefinitelyTyped. Do you think we should rely on them? They do not have typing for all options such as pluralRules but most of the other ones seem to be correct. I think it would be better to add types to the official node-polyglot package then maybe we could depend on them.

nayaabkhan commented on src/i18n.d.ts:
```diff
@@ -3,12 +3,37 @@
   children: ReactNode
   /** Locale to use, e.g. `en` */
-  locale: number
+  locale: string | number
```
Let's keep it only `string` if that is not a problem.

nayaabkhan commented on src/i18n.js:
```diff
@@ -38,5 +43,15 @@ export default class I18n extends React.Component {
 I18n.propTypes = {
   locale: PropTypes.string.isRequired,
   messages: PropTypes.object.isRequired,
+  allowMissing: PropTypes.bool,
```
Can you add `defaultProps` for the props that are not `isRequired`?

soda0289 replied:
Not sure we should set deafults for things like onMissingKey or pluralRules or interpolation. Unless we know some sane defaults to place in but I feel like node-polyglot should handle this.

nayaabkhan commented on src/i18n.d.ts:
```diff
@@ -3,12 +3,37 @@
+    [interpolationKey: string]: any;
+}
+
+interface InterpolationTokenOptions {
```
Same as https://github.com/nayaabkhan/react-polyglot/pull/24/files#r329338001 for this.

nayaabkhan commented on src/i18n.d.ts:
```diff
@@ -3,12 +3,37 @@
   /** A dictionary of translations */
   messages: object
+
+  allowMissing?: boolean,
```
I guess we could reuse these props rather in this way:

```ts
interface I18nProps extends Pick<PolyglotOptions, 'allowMissing' | 'onMissingKey' | 'interpolation'> {
  children: ReactNode
  /** A dictionary of translations */
  messages: object
  /** https://github.com/airbnb/polyglot.js#custom-pluralization-rules */
  pluralRules?: PluralRules
}
```

nayaabkhan commented:
@soda0289 @bastienrobert Thank you guys for your contribution, I have left my suggestions on the PR. Let me know what you think.

soda0289 commented:
@nayaabkhan Thanks for the review will push some fixes.

soda0289 commented:
@nayaabkhan Let me know if you think we should use the @types/node-polyglot definitions. I have left them out for now but can change it if you think its ok. I was also not sure what the default types should be for things like onMissingKey, interpolation, and pluralRules.
